tests/lax_scipy_special_functions_test.py

```diff
@@ -95,8 +95,11 @@ def op_record(name, nargs, dtypes, rng_factory, test_grad, nondiff_argnums=(), t
     op_record(
         "factorial", 1, float_dtypes, jtu.rand_default, True
     ),
+    # Cannot test tangent because float32 returns a constant value
+    # (i.e., very small but rapid oscillations for large x),
+    # but the derivative is non-zero.
     op_record(
-        "fresnel", 1, float_dtypes, jtu.rand_fresnel, True
+        "fresnel", 1, float_dtypes, jtu.rand_fresnel, False
```

Comment:
Yes, I know :/ (See my comment above)

Reply:
Incorrect gradients seems like a blocker for merging this – any ideas what might be going on?

Reply:
The issue is that, for single precision, large values of x results in a constant output: 0.5. Thus, the derivative computed with automatic diff is zero. However, the derivative is not zero in reality, and the custom JVP returns the correct derivative (that is trivial).

Reply:
Are you saying it’s a bug in JAX autodiff rather than a bug in your jvp rule?

Reply:
Not a bug, but a limitation.

The derivatives of Fresnel integrals are trivial: `cos(x * x * pi / 2)` and `sin(x * x * pi / 2)`. So, instead of relying on an estimate of the derivates through autodiff, I defined a `custom_jvp` function: https://github.com/google/jax/blob/9449e951b58e0d29673165d6cd10b3330454e729/jax/_src/third_party/scipy/special.py#L278-L287

This worked fine with the `float64` implementation of `fresnel`, as the autodiff value would be very close to that of `_fresnel_jvp`.

Now that I have added a special implementation for `float32`, the autodiff value is not close to the value returned by `_fresnel_jvp`. More precisely, when `x` is large, the following lines will return 0 with autodiff:
https://github.com/google/jax/blob/9449e951b58e0d29673165d6cd10b3330454e729/jax/_src/third_party/scipy/special.py#L239-L244

Which differs from the true derivative value (i.e., `_fresnel_jvp`).

Reply:
Note: it returns 0 because the output value is constant with respect to `x`.

Reply:
It looks like the problem stems from computing `x * x * pi / 2`. Your implementation avoids this by using `r = jnp.fmod(x, 2.0)` to get the correct phase without being subject to errors from float precision. You could probably use the same strategy in the case of the jvp rule.

Reply:
I will be off for one week, so I will look at that the week after :)

Reply:
Hey @jakevdp, so I will try to resume the problem we discussed above.
To recall, the problem is that `jtu.check_grads` fails because the estimated derivative with finite differences does not match the one provided with `custom_jvp`.

I don't think the problem comes from computing `x * x * pi / 2`, as you mentioned, but it *may* increase accuracy to use `sincospi(x)`.

Before b409c6394ea648ef7e9a9a341f859a998440ffd0 and 9449e951b58e0d29673165d6cd10b3330454e729, the tests for `check_grads` worked just fine. The aforementioned commits introduced a new branching for large, single-precision `x` values, where the output is approximately constant, i.e., `c_large = s_large = 0.5`. Hence, estimating the derivative of `fresnel(x)` when `x` is large will always return 0. For double-precision, `fresnel(x)` is not constant but oscillates very rapidly around `0.5`, with a small amplitude (see the above code snippet), so finite-difference may estimate the derivative correctly.

However, we know the derivative of `fresnel(x)` is easy to derive analytically, and this explains why I set up a custom VJP rule. This way, both single-precision and double-precision implementation of `fresnel(x)` have a very accurate estimate of their derivative(s).

As a result, I disabled the test for `check_grads`, but eventually it only needs to be disabled for `abs(x) > 36974.0 and (x.dtype == jnp.float32)`.

What do you think?

Reply:
> I don't think the problem comes from computing `x * x * pi / 2`

I do think this is the problem: for large `x`, `x * x` is even larger, and when it gets larger than about `10^7`, you can no longer resolve the difference between, say `1` and `1.5`, meaning that you entirely lose the phase information, which is the only relevant thing in the computation.

You can fix this by using modular arithmetic, bounded to the relevant domain of $0\le x < 2$

Reply:
Ok, I was wrong about my hypothesis, because setting `single_precision = False` should have solved the issue, in theory, but it did not. But now I understand better why you suggested the issue might be in capturing the phase :-)

I quickly tried your suggestion (see 5397fc29578726ca6a449aa2059dc40b650a0f8c), but that doesn't work for the moment (I was quick, so I probably made a mistake), to have `r = x**2 / 2 mod 2 = (x mod 2)(x/2 mod 2) mod 2`.

Reply:
Ok but the identity I used does not hold, I am dumb haha ^^'